src/graph/network.cc

```diff
@@ -78,104 +156,233 @@ DGL_REGISTER_GLOBAL("network._CAPI_DGLSenderConnect")
+    int port = args[2];
+    int num_sender = args[3];
+    network::Receiver* receiver = static_cast<network::SocketReceiver*>(chandle);
+    std::string addr;
+    if (receiver->Type() == "socket") {
+      addr = StringPrintf("socket://%s:%d", ip.c_str(), port);
+    } else {
+      LOG(FATAL) << "Unknown communicator type: " << receiver->Type();
+    }
+    if (receiver->Wait(addr.c_str(), num_sender) == false) {
+      LOG(FATAL) << "Wait sender socket failed.";
+    }
+  });
+
+
+////////////////////////// Distributed Sampler Components ////////////////////////////////
+
+
+DGL_REGISTER_GLOBAL("network._CAPI_SenderSendNodeFlow")
 .set_body([] (DGLArgs args, DGLRetValue* rv) {
     CommunicatorHandle chandle = args[0];
     int recv_id = args[1];
-    // TODO(minjie): could simply use NodeFlow nf = args[2];
     GraphRef g = args[2];
-    const IdArray node_mapping = args[3];
-    const IdArray edge_mapping = args[4];
-    const IdArray layer_offsets = args[5];
-    const IdArray flow_offsets = args[6];
+    const NDArray node_mapping = args[3];
+    const NDArray edge_mapping = args[4];
+    const NDArray layer_offsets = args[5];
+    const NDArray flow_offsets = args[6];
     auto ptr = std::dynamic_pointer_cast<ImmutableGraph>(g.sptr());
     CHECK(ptr) << "only immutable graph is allowed in send/recv";
-    network::Sender* sender = static_cast<network::Sender*>(chandle);
     auto csr = ptr->GetInCSR();
-    // Write control message
-    char* buffer = sender->GetBuffer();
-    *buffer = CONTROL_NODEFLOW;
-    // Serialize nodeflow to data buffer
-    int64_t data_size = network::SerializeSampledSubgraph(
-                             buffer+sizeof(CONTROL_NODEFLOW),
-                             csr,
-                             node_mapping,
-                             edge_mapping,
-                             layer_offsets,
-                             flow_offsets);
-    CHECK_GT(data_size, 0);
-    data_size += sizeof(CONTROL_NODEFLOW);
-    // Send msg via network
-    SendData(sender, buffer, data_size, recv_id);
+    // Create a message for the meta data of ndarray
+    MsgMeta msg(kNodeFlowMsg);
+    msg.AddArray(node_mapping);
+    msg.AddArray(edge_mapping);
+    msg.AddArray(layer_offsets);
+    msg.AddArray(flow_offsets);
+    msg.AddArray(csr->indptr());
+    msg.AddArray(csr->indices());
+    msg.AddArray(csr->edge_ids());
+    // send meta message
+    int64_t size = 0;
+    char* data = msg.Serialize(&size);
+    network::Sender* sender = static_cast<network::Sender*>(chandle);
+    Message send_msg;
+    send_msg.data = data;
+    send_msg.size = size;
+    send_msg.deallocator = DefaultMessageDeleter;
+    CHECK_NE(sender->Send(send_msg, recv_id), -1);
+    // send node_mapping
+    Message node_mapping_msg;
+    node_mapping_msg.data = static_cast<char*>(node_mapping->data);
+    node_mapping_msg.size = node_mapping.GetSize();
+    CHECK_NE(sender->Send(node_mapping_msg, recv_id), -1);
```

Comment:
pass the ownership of the memory in NDArray to sender. The simplest thing to do is to add the NDArray to the deallocator. @jermainewang is it the right way of avoiding the memory in NDArray from being free'd?

Reply:
@jermainewang When sending ndarray, we assume that user will use this NDArray in the future, so we don't set deallocator to the message here. If we want to clear reference here, how can we do it safely (avoid free the memory that user will use later)?

Reply:
Da is right. The simplest way to achieve this is to let the deallocator hold a reference to the NDArray. The deallocator body does nothing because when the message structure is destroyed, the deallocator function pointer is destroyed and so does the NDArray and its data.